packages/ui-components/src/components/PopupMenu/PopupMenu.component.tsx

```diff
@@ -0,0 +1,394 @@
+type HeadlessMenuItemsProps = React.ComponentPropsWithRef<React.ElementType>
+// Extend the extracted props instead:
+export interface PopupMenuMenuProps extends HeadlessMenuItemsProps {
+  as?: React.ElementType // Allow customizing the element type just as headless ui does
+  className?: string
+  children?: React.ReactNode
+  key?: React.Key
+}
+
+export interface PopupMenuItemProps extends React.ComponentProps<typeof HeadlessMenu.Item> {
+  as?: React.ElementType // Allow customising the element type to allow for rendering items as anchor elements, just as headless ui does
+  className?: string
+  disabled?: boolean
+  href?: string // accept a href when rendering a link, whether or not to pass it to HUI item will behandled in the component
+  icon?: keyof typeof KnownIconsEnum
+  label?: string
+  rel?: string // accept rel for anchors, handle in component
+  target?: string // accepot target for links, handle in component
+}
+
+export interface PopupMenuSectionProps {
+  children?: React.ReactNode
+  className?: string
+}
+
+export interface PopupMenuSectionHeadingProps {
+  children?: React.ReactNode
+  className?: string
+  label?: string
+}
+
+export interface PopupMenuSectionSeparatorProps {
+  className?: string
+}
+
+// ----- Define PopupMenu context and hook -----
+
+// Context:
+const PopupMenuContext = createContext<PopupMenuContextType | null>(null)
+
+// Hook to use the context:
+export const usePopupMenuContext = () => {
+  const context = useContext(PopupMenuContext)
+  if (!context) {
+    throw new Error(
+      "The usePopupMenuContext hook must be used inside within a PopupMenu that provides the PopupMenuContext."
+    )
+  }
+  return context
+}
+
+// Export the context for advanced use cases:
+export { PopupMenuContext }
+
+// ----- Component Definitions -----
+
+// POPUP MENU
+
+/** A Popup Menu component that wraps Headless UI Menu. The Menu will be rendered into a Juno Portal, so using Juno's PortalProvider (which is already included when using Juno's AppShell) is mandatory. */
+const PopupMenu: React.FC<PopupMenuProps> & {
+  Toggle: React.FC<PopupMenuToggleProps>
+  Menu: typeof PopupMenuMenu
+  Item: React.FC<PopupMenuItemProps>
+  Section: React.FC<PopupMenuSectionProps>
+  SectionHeading: React.FC<PopupMenuSectionHeadingProps>
+  SectionSeparator: React.FC<PopupMenuSectionSeparatorProps>
+} = ({
+  children = null,
+  className = "",
+  disabled = false,
+  icon = "moreVert",
+  menuSize = "normal",
+  onClose = undefined,
+  onOpen = undefined,
+  ...props
+}) => {
+  // Create a state to track headless-ui's internal open state (Note that we can track the internal state and can let headless ui update our tracking state, but we can not set it, hence there is no `open` prop on our component):
+  const [isOpen, setIsOpen] = useState(false)
+  const previousOpenRef = useRef(isOpen) // Track the previous value of isOpen for efficiency
```

Comment:
could you clarify why you need this ?

Reply:
Headless Menu does not directly expose a state whether it is open or not, the only thing we can get is a render prop. In order to run our handlers and overall keep our APIs consistent as much as possible, I have introduced a state that tracks the internal state of Headless Menu, so we can read, add it to our context so we and users of our components can use it whenever and however they/we want. It is updated during the render cycle when Headless Menu's `open` render prop changes.